README.md
Chef-Bareos Cookbook
====================

[![BuildStatus](https://travis-ci.org/sitle/chef-bareos.svg?branch=master)](https://travis-ci.org/sitle/chef-bareos)
[![Join the chat at https://gitter.im/EMSL-MSC/chef-bareos](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/EMSL-MSC/chef-bareos?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)  

This cookbook installs and configures backups based on [BAREOS](https://www.bareos.org/en/).

[Official BAREOS Documentation](http://doc.bareos.org/master/html/bareos-manual-main-reference.html).

## Supported Platforms:
 * Ubuntu 14.04 (plan to add 16.04 as soon as binary is released)
 * Debian 7 (8+ may or may not work, you'll need a repo basically)
 * CentOS 6+
 * RHEL 6+ (Assumed to work just as well as on CentOS)

## Supported Chef Versions:

 * chef-client '>= 12.9.38'
 * chef-server '>= 12.5'

## Important Notable Attributes

### Repository
Assists with adding necessary sources for installing Bareos

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['url'] | 'http://download.bareos.org/bareos/release' | Main installation URL
| ['bareos']['contrib_url'] | 'http://download.bareos.org/bareos/contrib' | Main contrib installation URL
| ['bareos']['version'] | '15.2' | Default Bareos Version

For other platform specific attributes please see default attributes file for more detail.

### Messages
Defines default admin e-mail address for service notifications and what messages to care about

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['messages']['mail_to'] | "bareos@#{node['domain']}" | Default messages e-mail destination
| ['bareos']['messages']['default_messages'] | 'Standard' | Default client message capture level
| ['bareos']['messages']['default_admin_messages'] | 'all, !skipped, !restored' | Default server message capture level
### Database
Populates the Catalog resource in the main Director configuration

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['database']['catalog_name'] | 'MyCatalog' | Default catalog name
| ['bareos']['database']['database_type'] | 'postgresql' | Default database installment indicator
| ['bareos']['database']['dbdriver'] | 'postgresql' | Config entry for type of database
| ['bareos']['database']['dbname'] | 'bareos' | Default database name
| ['bareos']['database']['dbuser'] | 'bareos' | Default database user
| ['bareos']['database']['dbpassword'] | *blank string* | Default gets generated by postgresql cookbook, unless specified here
| ['bareos']['database']['dbaddress'] | nil | This is often just the localhost, so this is only needed in certain cases

### Clients
Provides resources for the Catalog (Director configuration) and Filedaemon configurations/templates

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['clients']['name'] | node['fqdn'] | Uses the node FQDN by default as prefix for filedaemon name
| ['bareos']['clients']['net_name'] | node['fqdn'] | DNS Network name used to resolve client
| ['bareos']['clients']['fd_port'] | 9102 | Default filedaemon port
| ['bareos']['clients']['max_concurrent_jobs'] | 20 | Default number of concurrent jobs
| ['bareos']['clients']['heartbeat_interval'] | 600 | Proven a useful default value, change as needed
| ['bareos']['clients']['client_search_query'] | 'roles:bareos_client' | Default search query to find Bareos clients
| ['bareos']['clients']['client_list'] | %w(node) | Useful if you need a list of hosts if running in solo mode
| ['bareos']['clients']['bootstrap_file'] | '/var/lib/bareos/%c.bsr' | Default bootstrap file structure/location
| ['bareos']['clients']['jobdef_default_messages'] | 'Standard' | Default value for setting the message level in a job definition to override the messages section
| ['bareos']['clients']['jobdef_default_fileset'] | "#{node['fqdn']}-Fileset" | Default naming convention for filesets
| ['bareos']['clients']['storage'] | 'File' | Default storage for new clients when added via search
| ['bareos']['clients']['sensitive_configs'] | true | Enables or disables sensitive resource configuration files

### Storage Daemon
Provides for a baseline Storage Daemon Config with configurable options

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['storage']['name'] | node['fqdn'] | Uses FQDN for naming storages found via search
| ['bareos']['storage']['storage_search_query'] | 'roles:bareos_storage' | Default search query string for finding storage servers
| ['bareos']['storage']['sd_port'] | 9103 | Default Storage communication port
| ['bareos']['storage']['servers'] | %w(node) | List of storage servers you can use if using solo mode
| ['bareos']['storage']['max_concurrent_jobs'] | 20 | Default max number of concurrent storage daemon jobs
| ['bareos']['storage']['autochanger_enabled'] | false | Used to control autochanger support
| ['bareos']['storage']['sensitive_configs'] | true | Enables or disables sensitive resource configuration files

### Director
Provides standard variables for a typical Director configuration

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['director']['name'] | node['fqdn'] | Uses FQDN for director naming
| ['bareos']['director']['net_name'] | node['fqdn'] | Uses FQDN for DNS resolution
| ['bareos']['director']['dir_search_query'] | 'roles:bareos_director' | Default search string to find bareos directors
| ['bareos']['director']['dir_port'] | 9101 | Default director communication port
| ['bareos']['director']['dir_max_concurrent_jobs'] | 20 | Default max allowable jobs running
| ['bareos']['director']['servers'] | %w(node) | List of directors if running in solo mode
| ['bareos']['director']['console_commandacl'] | 'status, .status' | Default ACL for console commands
| ['bareos']['director']['heartbeat_interval'] | 600 | Proven useful as a default network timeout for communication to director
| ['bareos']['director']['catalog_jobdef'] | 'default-catalog-def' | Default name for the Catalog Backup Jobdef name
| ['bareos']['director']['conf']['help']['Example Block'] | '# You can put extra configs here.' | Area where you can add any number of possible things to expand your configs
| ['bareos']['director']['config_change_notify'] | 'restart' | Default action when director config changes (restart/reload)
| ['bareos']['director']['sensitive_configs'] | true | Enables or disables sensitive resource configuration files


### Subscription Management (Director)
Provides a system counter method if you have a paid service subscription

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['director']['dir_subscription'] | nil | Required if you have a support contract/licensed installation (activates if not nil)
| ['bareos']['director']['dir_subs'] | nil | Max number of subs you have signed up for

### Workstation
Determines if you want to use FQDN or some other way of defining hosts in your management workstation deployment

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['workstation']['name'] | node['fqdn'] | Used to determine header information for bconsole/bat configs
| ['bareos']['workstation']['sensitive_configs'] | true | Enables or disables sensitive resource configuration files

### Graphite Plugin
A new plugin that will send statistics to a graphite server which can then be used in various ways.

| Attribute        | Default Value | Description
|------------------|---------------|------------
| ['bareos']['plugins']['graphite']['plugin_path'] | '/usr/sbin' | Default location for the plugin that runs in a defined cron job
| ['bareos']['plugins']['graphite']['config_path'] | '/etc/bareos' | Default directory for the plugin config
| ['bareos']['plugins']['graphite']['search_query'] | 'roles:bareos_director' | Default search string to populate the director name
| ['bareos']['plugins']['graphite']['server'] | 'graphite' | Placeholder string for the graphite server DNS name
| ['bareos']['plugins']['graphite']['graphite_port'] | '2003' | Default graphite communication port
| ['bareos']['plugins']['graphite']['graphite_data_prefix'] | 'bareos.' | Default prefix for graphite data
| ['bareos']['plugins']['graphite']['graphite_plugin_src_url'] | See attributes file | Default URL to the plugin
| ['bareos']['plugins']['graphite']['cron_job'] | nil | Activates a general minutely cronjob if defined other than nil
| ['bareos']['plugins']['graphite']['cron_command'] | See attribute file | Default cron script to run for the graphite plugin
| ['bareos']['plugins']['sensitive_configs'] | true | Enables or disables sensitive resource configuration files

## Recipes

### default
Installs the Bareos repos (via the repo recipe) and the client filedaemon (via the client recipe). Please NOTE, although it will install these parts, the director will not pick up on the client unless you (by default) create and attach a bareos_client role to hosts you wish to configure automatically. You can also add the host to the director in the unmanaged-host key value hashes.

### client
Installs the Bareos filedaemon and creates a config file that is linked to available directors on chef server.
You may also feed directors to the config via attributes if running in solo mode.

### repo
Installs base Bareos repo as well as the Bareos Contrib repo.

### database
Installs whichever database is desired per attributes (PostgreSQL/MySQL), installs Bareos database packages and creates the bareos database and user for you. Should also set the database password by default. You may need to recover this from the attributes or set a new one via vault via wrapper recipe.

### server
Installs necessary Bareos server packages and sets up base configs necessary for server to start. Also creates the config directory (bareos-dir.d) so you can drop whatever outside config files into place and have them get automatically included in your setup.

### storage
Installs necessary Bareos storage packages and sets up a default file storage for you to start backing stuff up to right away (configured for ~250GB spread over 25 10GB volumes).

### autochanger
This bit will setup an autochanger based on a pretty straight forward has table. Tested with IBM TS3500 Tape Library with 10 Frames and 16 Tape drives.

### workstation
Installs the bconsole utility. There are future plans to create a recipe to install bat (Bareos Administration Tool) and the Bareos Web UI.

### graphite_plugin
Installs a Bareos graphite plugin, configuration file, necessary python packages, and a cronjob to gather statistics periodically and forward them to an available graphite server.

## Searchable Roles (Used by default)

### bareos\_client
This example shows how the ```bareos_client``` role can both install the Bareos client side software and when searched against via the server recipe, will add itself to the bareos-dir (Bareos director) configuration and setup a default set of jobs for a client.

```
{
  "name": "bareos_client",
  "description": "Example Role for Bareos clients using the chef-bareos Cookbook, used in searches, throws down sources for installs",
  "json_class": "Chef::Role",
  "default_attributes": {
  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "recipe[chef-bareos]"
  ],
  "env_run_lists": {
  }
}
```

### bareos\_storage
This example shows a ```bareos_storage``` role which will create a Bareos storage-daemon host. It will install the necessary packages and lay down configuration files you can populate with any number of key value hash tables. You should be able to install this independent of the director(s), please file a ticket if this doesn't work as expected.

```
{
  "name": "bareos_storage",
  "description": "Example Role for a Bareos storage",
  "json_class": "Chef::Role",
  "default_attributes": {
  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "recipe[chef-bareos::storage]"
  ],
  "env_run_lists": {
  }
}
```

### bareos\_director
This role example will install an all-in-one Bareos director server. This includes a "first" client, bareos database, bareos storage, and deploy bconsole to interact with the director.

This will also allow clients to populate their filedaemon config via defined search string.

You'll need to run ```chef-client``` on the director after a client gets configured so the director can add and generate the appropriate client related configs.

You can populate the ```['bareos']['clients']['unmanaged']``` hash table space with any number of client related configuration lines if you have hosts you either don't plan to search for or want to do custom configurations for.

```
{
  "name": "bareos_director",
  "description": "Example Role for a Bareos director",
  "json_class": "Chef::Role",
  "default_attributes": {
  },
  "override_attributes": {
  },
  "chef_type": "role",
  "run_list": [
    "role[bareos_client]",
    "recipe[chef-bareos::database]",
    "recipe[chef-bareos::server]",
    "recipe[chef-bareos::workstation]"
  ],
  "env_run_lists": {
  }
}
```

## Example customizable key value hash template configurations
These are the preset default hashes to get a baseline configuration on a new bareos server. You can manipulate these as you see fit via recipe logic or searches or whatever you want. These will at least get you going.

### clients
```
# Default Client Config populated via search
default['bareos']['clients']['conf'] = {
  'FDPort' => '9102',
  'File Retention' => '30 days',
  'Job Retention' => '6 months',
  'AutoPrune' => 'yes',
  'Maximum Concurrent Jobs' => '20'
}
```

```
# Example Unmanaged client if client is unmanaged or custom
default['bareos']['clients']['unmanaged']['unmanaged-client-fd'] = {
  'Address' => 'unmanaged-client',
  'Password' => 'onefbofnerwob',
  'Catalog' => 'MyCatalog',
  'FDPort' => '9102'
}
```

### autochanger (if using tape storage)
```
# Example/Test Tape Autochanger Configurations
if node['bareos']['storage']['autochanger_enabled'] == true
  default['bareos']['storage']['autochangers']['autochanger-0'] = {
    'Device' => [
      'tapedrive-0',
      'tapedrive-1'
    ],
    'Changer Device' => ['/dev/tape/by-id/scsi-1TANDBERGStorageLoader_SOMEAUTOCHANGER'],
    'Changer Command' => ['"/usr/lib/bareos/scripts/mtx-changer %c %o %S %a %d"']
  }

  default['bareos']['storage']['autochangers']['autochanger-1'] = {
    'Device' => [
      'tapedrive-0'
    ],
    'Changer Device' => ['/dev/tape/by-id/scsi-1TANDBERGStorageLoader_SOMEAUTOCHANGER'],
    'Changer Command' => ['"/usr/lib/bareos/scripts/mtx-changer %c %o %S %a %d"']
  }

  default['bareos']['storage']['devices']['tapedrive-0'] = {
    'DeviceType' => 'tape',
    'DriveIndex' => '0',
    'ArchiveDevice' => 'dev/nst0',
    'MediaType' => 'lto',
    'Autochanger' => 'no',
    'AutomaticMount' => 'no',
    'MaximumFileSize' => '10GB'
  }

  default['bareos']['storage']['devices']['tapedrive-1'] = {
    'DeviceType' => 'tape',
    'DriveIndex' => '0',
    'ArchiveDevice' => 'dev/nst0',
    'MediaType' => 'lto',
    'Autochanger' => 'no',
    'AutomaticMount' => 'no',
    'MaximumFileSize' => '10GB'
  }
```

### dir\_helper
```
default['bareos']['director']['conf']['help']['Example Block'] = '# You can put extra configs here.'
```

### filesets
```
# Default Filesets
default['bareos']['clients']['filesets']['default'] = {
  'options' => {
    'signature' => 'MD5'
  },
  'include' => {
    'File' => ['/', '/home'],
    'Exclude Dir Containing' => ['.bareos_ignore']
  },
  'exclude' => {

CHANGELOG.md
CHANGELOG
=========
Chef-Bareos Cookbook
--------------------

3.0.8
-----
#### Ian Smith
  * Updating the cronjob for the graphite plugin so it is less noisy if something fails by default
  * Adding a default attribute to control the exact cronjob script the user wants
  * Fixing up the Chefspec tests for the graphite cronjob resource

3.0.6
-----
#### Ian Smith
  * Validating compatibility with chef-client 12.14.89
  * Updated kitchen config for chef-client 12.14.89
  * Dropping support for ruby <2.2.0 for TravisCI tests

3.0.5
-----
#### Ian Smith
  * Adding an attribute for each major config template so admins can enable or disable sensitive resources. Before it was hard coded to be sensitive
  * Discovered a bug that basically broke bconsole because the password was being regenerated on every chef run which requires a restart not a reload. Changed to node.normal from node.default

3.0.4
-----
#### Ian Smith
  * Updating the cronjob resource to only be included if user specifies a desire to enable a cronjob, defaults to off

3.0.3
-----
#### Ian Smith
  * Decided to change who owned all of the graphite_plugin bits
    1. Bareos plugin still seems to have some debugging code that will not only PRINT your password to the command line if done by hand, but will also EMAIL your main director password to whatever account owns, say, a cronjob (i.e. root). Seems like a security flaw? I submitted bugfix bareos/bareos-contrib/#14 to address this issue, waiting for merge. If you need an immediate fixed source, use my GitBytes/bareos-contrib fork. Sorry about that..
    1. Pretty much wherever root was, I changed to bareos in the plugin recipe.
  * Cleanup README and CHANGELOG raw syntax for GitHub and Supermarket rendering
  * Add new chefspec/rspec unit tests, cleanup existing tests
    1. Removed unused `.rspec` config file
    1. Removed `repo_spec` test, never quite worked right
    1. Added `:documentation` and `color` output to chefspec/rspec testing
    1. Added `graphite_plugin_spec` test
    1. Condensed and cleaned up `default_spec` test, tests both the `repo` and `client` recipes on top of the implied `default` recipe
    1. Added ability to easily run tests against a variety of preset supported platforms via `spec/unit/support/supported_platforms` definition file (see `default_spec` test for example usage)
  * Add new/optimize serverspec integration tests
    1. Condense tests for overall functionality of Bareos
    1. New graphite plugin testing: plugin, config, cronjob
  * Bump patch version to `3.0.3` (tag v3.0.3)

3.0.2
-----
#### Ian Smith
  * Complete re-work of cookbook, which includes the rework bits from version 2.0
  * Adding various features including:
    * Templates populated from hash tables
    * Up to date README
    * Graphite plugin deployment recipe
    * Migrate to postgresql cookbook version 4+
    * Better testing (rspec), testing against current ruby, need some more rspec tests but no time
    * Various other updates and enhancements, see README for details
    * Supermarket Release...finally
  * ***NOTE*** This release adds functionality that is not backwards compatible. Version lock until you have time to test the migration to version 3.0+

2.2.13
------
#### Ian Smith
  * Near complete refactoring of how the cookbook works
    * Adding various ways to add configs based on hashes. Examples in attributes/default.rb
  * Updating README for usage details
  * If you are not prepared to use this cookbook version, please lock cookbook version to < 2.0.0
  * TODO: Refactor main bareos-dir/sd/fd configs to be fully based on hashes like other configs
    * Remaining attributes support primary bareos-dir/sd/fd configs mostly, used elsewhere but not much

1.0.4
-----
#### Ian Smith
  * Updating and validating Kitchen testing with new methods
  * Adding ability to use custom Storage Daemon Configs via wrapper recipe
  * Fixed a race condition that was likely to come up both in testing and production
    * Was missing client config on fresh install so restart of dir would fail
  * Limiting cookbook support for debian to ~>7.0 until we can test on 8+
  * Refined storage recipe to correctly create client configs based on BOTH role search and solo instances, may need a bit more tuning but should work in most cases
  * Creating basic Full, Inc, and Diff pools even if not used, just as a place holder (default)
  * Minor fixes
  * Larger version bump for version clarity
  * Updated Docs
  * Adding support for Fedora and RHEL 6/7
  * Reworked the available repos to some better logic

0.1.4
-----
#### Ian Smith
  * Updating README
  * Revamping server.rb recipe to better utilize the bareos-dir.d directory
    * Moving ALL host config files to the bareos-dir.d/hosts/ directory
    * Revamping director config and moving host bits to separate host.conf.erb template
    * Adding feature for host based custom pools based on true/false attribute
    * Adding ability to fill in a block for virtual-full backup logic (not complete yet)
  * Revamping storage.rb and storage template to what I am thinking was meant to happen originally
  * Adding new kitchen suite and tests to verify host pools work
  * Bugfixes
  * Updating Kitchen Tests and ChefSpec Configs to match for verifying
    * Addressing race condition in ChefSpec tests where debian was taking
      longer than expected to start the director

0.1.3
-----
#### Ian Smith
  * Updating README
  * Updating CHANGELOG
  * Updating metadata file

0.1.2
-----
#### Leonard TAVAE
  * The License has changed (Apache 2.0)
  * The cookbook now passed foodcritic, rubocop and tailor with success
  * Some minor bugs fix

#### Ian Smith
  * Updating Cookbook to use ChefDK >0.6.0
  * Updating Docs
  * Updating Postgresql Logic/Commands
  * Updating a couple of template bits for the director
    * Adding new attributes, working on more for future implementation
  * Updating Kitchen suites/platforms/provisioner
    * Adding some test roles for tests

attributes/default.rb
# Repository
default['bareos']['url'] = 'http://download.bareos.org/bareos/release'
default['bareos']['contrib_url'] = 'http://download.bareos.org/bareos/contrib'

default['bareos']['version'] = '15.2' # <--- Latest Stable version as of 06-Dec-2015

if platform_family?('rhel', 'fedora')
  default['bareos']['repository_name'] = 'bareos'
  default['bareos']['description'] = "Backup Archiving REcovery Open Sourced Current #{node['bareos']['version']}"
  default['bareos']['contrib_repository_name'] = 'bareos_contrib'
  default['bareos']['contrib_description'] = "Backup Archiving REcovery Open Sourced Current #{node['bareos']['version']} contrib"
end

case node['platform_family']
when 'debian'
  case node['platform']
  when 'debian'
    default['bareos']['baseurl'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/Debian_#{node['platform_version'].to_i}.0/"
    default['bareos']['gpgkey'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/Debian_#{node['platform_version'].to_i}.0/Release.key"
    default['bareos']['contrib_baseurl'] = "#{node['bareos']['contrib_url']}/Debian_#{node['platform_version'].to_i}.0/"
    default['bareos']['contrib_gpgkey'] = "#{node['bareos']['contrib_url']}/Debian_#{node['platform_version'].to_i}.0/Release.key"
  when 'ubuntu'
    default['bareos']['baseurl'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/xUbuntu_#{node['platform_version']}/"
    default['bareos']['gpgkey'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/xUbuntu_#{node['platform_version']}/Release.key"
    default['bareos']['contrib_baseurl'] = "#{node['bareos']['contrib_url']}/xUbuntu_#{node['platform_version']}/"
    default['bareos']['contrib_gpgkey'] = "#{node['bareos']['contrib_url']}/xUbuntu_#{node['platform_version']}/Release.key"
  end
when 'rhel'
  case node['platform']
  when 'centos'
    default['bareos']['baseurl'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/CentOS_#{node['platform_version'].to_i}/"
    default['bareos']['gpgkey'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/CentOS_#{node['platform_version'].to_i}/repodata/repomd.xml.key"
    default['bareos']['contrib_baseurl'] = "#{node['bareos']['contrib_url']}/CentOS_#{node['platform_version'].to_i}/"
    default['bareos']['contrib_gpgkey'] = "#{node['bareos']['contrib_url']}/CentOS_#{node['platform_version'].to_i}/repodata/repomd.xml.key"
  when 'scientific'
    default['bareos']['baseurl'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/CentOS_#{node['platform_version'].to_i}/"
    default['bareos']['gpgkey'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/CentOS_#{node['platform_version'].to_i}/repodata/repomd.xml.key"
    default['bareos']['contrib_baseurl'] = "#{node['bareos']['contrib_url']}/CentOS_#{node['platform_version'].to_i}/"
    default['bareos']['contrib_gpgkey'] = "#{node['bareos']['contrib_url']}/CentOS_#{node['platform_version'].to_i}/repodata/repomd.xml.key"
  when 'redhat'
    default['bareos']['baseurl'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/#{node['platform_family'].upcase}_#{node['platform_version'].to_i}/"
    default['bareos']['gpgkey'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/#{node['platform_family'].upcase}_#{node['platform_version'].to_i}/repodata/repomd.xml.key"
    default['bareos']['contrib_baseurl'] = "#{node['bareos']['contrib_url']}/#{node['platform_family'].upcase}_#{node['platform_version'].to_i}/"
    default['bareos']['contrib_gpgkey'] = "#{node['bareos']['contrib_url']}/#{node['platform_family'].upcase}_#{node['platform_version'].to_i}/repodata/repomd.xml.key"
  end
when 'fedora'
  default['bareos']['baseurl'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/#{node['platform'].capitalize}_#{node['platform_version']}/"
  default['bareos']['gpgkey'] = "#{node['bareos']['url']}/#{node['bareos']['version']}/#{node['platform'].capitalize}_#{node['platform_version']}/repodata/repomd.xml.key"
  default['bareos']['contrib_baseurl'] = "#{node['bareos']['contrib_url']}/#{node['platform'].capitalize}_#{node['platform_version']}/"
  default['bareos']['contrib_gpgkey'] = "#{node['bareos']['contrib_url']}/#{node['platform'].capitalize}_#{node['platform_version']}/repodata/repomd.xml.key"
else
  default['bareos']['baseurl'] = nil
  default['bareos']['gpgkey'] = nil
  default['bareos']['contrib_baseurl'] = nil
  default['bareos']['contrib_gpgkey'] = nil
end

# Messages
default['bareos']['messages']['mail_to'] = "bareos@#{node['domain']}"
default['bareos']['messages']['default_messages'] = 'Standard'
default['bareos']['messages']['default_admin_messages'] = 'all, !skipped, !restored'

# Database
default['bareos']['database']['catalog_name'] = 'MyCatalog'
default['bareos']['database']['database_type'] = 'postgresql' # postgresql/mysql
default['bareos']['database']['dbdriver'] = 'postgresql' # postgresql/mysql
default['bareos']['database']['dbname'] = 'bareos'
default['bareos']['database']['dbuser'] = 'bareos'
default['bareos']['database']['dbpassword'] = ''
default['bareos']['database']['dbaddress'] = nil

# Clients
default['bareos']['clients']['name'] = node['fqdn']
default['bareos']['clients']['net_name'] = node['fqdn']
default['bareos']['clients']['fd_port'] = 9102
default['bareos']['clients']['max_concurrent_jobs'] = 20
default['bareos']['clients']['heartbeat_interval'] = 600
default['bareos']['clients']['client_search_query'] = 'roles:bareos_client'
default['bareos']['clients']['client_list'] = %w(node)
default['bareos']['clients']['bootstrap_file'] = '/var/lib/bareos/%c.bsr'
default['bareos']['clients']['jobdef_default_messages'] = 'Standard'
default['bareos']['clients']['jobdef_default_fileset'] = "#{node['fqdn']}-Fileset"
default['bareos']['clients']['storage'] = 'File'
default['bareos']['clients']['sensitive_configs'] = true

# Storage Daemon
default['bareos']['storage']['name'] = node['fqdn']
default['bareos']['storage']['storage_search_query'] = 'roles:bareos_storage'
default['bareos']['storage']['sd_port'] = 9103
default['bareos']['storage']['servers'] = %w(node)
default['bareos']['storage']['max_concurrent_jobs'] = 20
default['bareos']['storage']['sd_mon_enable'] = 'yes'
default['bareos']['storage']['autochanger_enabled'] = false
default['bareos']['storage']['conf']['help']['Example Block'] = '# You can put extra configs here.'
default['bareos']['storage']['sensitive_configs'] = true

# Director
default['bareos']['director']['name'] = node['fqdn']
default['bareos']['director']['net_name'] = node['fqdn']
default['bareos']['director']['dir_search_query'] = 'roles:bareos_director'
default['bareos']['director']['dir_port'] = 9101
default['bareos']['director']['dir_max_concurrent_jobs'] = 20
default['bareos']['director']['servers'] = %w(node)
default['bareos']['director']['console_commandacl'] = 'status, .status'
default['bareos']['director']['heartbeat_interval'] = 600
default['bareos']['director']['catalog_jobdef'] = 'default-catalog-def'
default['bareos']['director']['conf']['help']['Example Block'] = '# You can put extra configs here.'
default['bareos']['director']['config_change_notify'] = 'restart'
default['bareos']['director']['sensitive_configs'] = true

# Subscription Management (Director)
default['bareos']['director']['dir_subscription'] = nil
default['bareos']['director']['dir_subs'] = nil

# Workstation
default['bareos']['workstation']['name'] = node['fqdn']
default['bareos']['workstation']['sensitive_configs'] = true

# Graphite Plugin Default Attributes
default['bareos']['plugins']['graphite']['plugin_path'] = '/usr/sbin'
default['bareos']['plugins']['graphite']['config_path'] = '/etc/bareos'
default['bareos']['plugins']['graphite']['search_query'] = 'roles:bareos_director'
default['bareos']['plugins']['graphite']['server'] = 'graphite'
default['bareos']['plugins']['graphite']['graphite_port'] = '2003'
default['bareos']['plugins']['graphite']['graphite_data_prefix'] = 'bareos.'
default['bareos']['plugins']['graphite']['graphite_plugin_src_url'] = 'https://raw.githubusercontent.com/bareos/bareos-contrib/master/misc/performance/graphite/bareos-graphite-poller.py'
default['bareos']['plugins']['graphite']['mail_to'] = ''
default['bareos']['plugins']['graphite']['sensitive_configs'] = true
default['bareos']['plugins']['graphite']['cron_command'] = "#{node['bareos']['plugins']['graphite']['plugin_path']}/bareos-graphite-poller.py -c #{node['bareos']['plugins']['graphite']['config_path']}/graphite-poller.conf >/dev/null 2>&1"

##############################
# Examples -  Default Hashes #
##############################

# Default Client Config
default['bareos']['clients']['conf'] = {
  'FDPort' => '9102',
  'File Retention' => '30 days',
  'Job Retention' => '6 months',
  'AutoPrune' => 'yes',
  'Maximum Concurrent Jobs' => '20'
}

default['bareos']['director']['jobs'] = nil

# Default Job Definitions
default['bareos']['clients']['job_definitions']['default-def'] = {
  'Level' => 'Incremental',
  'Fileset' => 'default-fileset',
  'Schedule' => 'monthly',
  'Storage' => 'default-file-storage',
  'Messages' => 'Standard',
  'Pool' => 'default-file-pool',
  'Priority' => '10',
  'Write Bootstrap' => '"/var/lib/bareos/%c.bsr"',
  'SpoolData' => 'no'
}

default['bareos']['clients']['job_definitions']['default-catalog-def'] = {
  'Level' => 'Full',
  'Fileset' => 'Catalog',
  'Schedule' => 'WeeklyCycleAfterBackup',
  'Storage' => 'default-file-storage',
  'Messages' => 'Standard',
  'Pool' => 'default-file-pool',
  'Allow Duplicate Jobs' => 'no'
}

default['bareos']['clients']['job_definitions']['default-restore-def'] = {
  'Fileset' => 'default-fileset',
  'Storage' => 'default-file-storage',
  'Messages' => 'Standard',
  'Pool' => 'default-file-pool',
  'Priority' => '7',
  'Where' => '/tmp/bareos-restores'
}

# Default Filesets
default['bareos']['clients']['filesets']['default'] = {
  'options' => {
    'signature' => 'MD5'
  },
  'include' => {
    'File' => ['/', '/home'],
    'Exclude Dir Containing' => ['.bareos_ignore']
  },
  'exclude' => {
    'File' => [
      '/var/lib/bareos',
      '/var/lib/bareos/storage',
      '/var/lib/pgsql',
      '/var/lib/mysql',
      '/proc',
      'tmp',
      '/.journal',
      '/.fsck',
      '/spool'
    ]
  }
}

# Default Pools
default['bareos']['clients']['pools']['default-file-pool'] = {
  'Pool Type' => 'Backup',
  'Recycle' => 'yes',
  'Volume Retention' => '30 days',
  'Maximum Volume Bytes' => '10G',
  'Maximum Volumes' => '25',
  'LabelFormat' => 'FileVolume-'
}

# Default Schedules
default['bareos']['clients']['schedules']['monthly'] = {
  'Description' => [
    'Default Monthly Schedule'
  ],
  'Run' => [
    'Full 1st sun at 23:05',
    'Differential 2nd-5th sun at 23:05',
    'Incremental mon-sat at 23:05'
  ],
  'Enabled' => [
    'yes'
  ]
}

# Default Storages
default['bareos']['clients']['storages']['default-file-storage'] = {
  'Address' => node['bareos']['storage']['name'], # N.B. Use a fully qualified name here
  'Device' => 'FileStorage',
  'Media Type' => 'File'
}

# Example/Test Tape Autochanger Configurations
if node['bareos']['test_mode'] == true && node['bareos']['storage']['autochanger_enabled'] == true
  default['bareos']['storage']['autochangers']['autochanger-0'] = {
    'Device' => [
      'tapedrive-0',
      'tapedrive-1'
    ],
    'Changer Device' => ['/dev/tape/by-id/scsi-1TANDBERGStorageLoader_SOMEAUTOCHANGER'],
    'Changer Command' => ['"/usr/lib/bareos/scripts/mtx-changer %c %o %S %a %d"']
  }

  default['bareos']['storage']['autochangers']['autochanger-1'] = {
    'Device' => [
      'tapedrive-0'
    ],
    'Changer Device' => ['/dev/tape/by-id/scsi-1TANDBERGStorageLoader_SOMEAUTOCHANGER'],
    'Changer Command' => ['"/usr/lib/bareos/scripts/mtx-changer %c %o %S %a %d"']
  }

  default['bareos']['storage']['devices']['tapedrive-0'] = {
    'DeviceType' => 'tape',
    'DriveIndex' => '0',
    'ArchiveDevice' => 'dev/nst0',
    'MediaType' => 'lto',
    'Autochanger' => 'no',
    'AutomaticMount' => 'no',
    'MaximumFileSize' => '10GB'
  }

  default['bareos']['storage']['devices']['tapedrive-1'] = {
    'DeviceType' => 'tape',
    'DriveIndex' => '0',
    'ArchiveDevice' => 'dev/nst0',
    'MediaType' => 'lto',
    'Autochanger' => 'no',
    'AutomaticMount' => 'no',
    'MaximumFileSize' => '10GB'
  }

  # Example Unmanaged client for testing
  default['bareos']['clients']['unmanaged']['unmanaged-client-fd'] = {
    'Address' => 'unmanaged-client',
    'Password' => 'onefbofnerwob',
    'Catalog' => 'MyCatalog',
    'FDPort' => '9102'
  }
  # Example Jobs
  default['bareos']['clients']['jobs']["#{node.default['bareos']['clients']['name']}-job"] = {
    'Client' => "#{node['bareos']['clients']['name']}-fd",
    'Type' => 'Backup',
    'JobDefs' => 'default-def'
  }

  default['bareos']['clients']['jobs']["#{node.default['bareos']['clients']['name']}-restore-job"] = {
    'Client' => "#{node['bareos']['clients']['name']}-fd",
    'Type' => 'Restore',
    'JobDefs' => 'default-restore-def'
  }
end
